### ISSUE TITLE
Rule to prevent no-op promise handlers

### ISSUE THREAD

**Issue by ForbesLindesay** [eslint/eslint#5761]
**When does this rule warn?**

When the identify function is used as the success handler of a Promise or the "throw function" is used as the failure handler of a Promise.

```js
// The identity function is the default success handler for a promise, it can be ommitted
something.then(function (value) {
  return value;
});

// and

// propagating the error is the default error handler fro a promise, it can be ommitted

something.then(null, function (err) {
  throw err;
});

// or

something.catch(function (err) {
  throw err;
});
```

**Is this rule preventing an error or is it stylistic?**

It helps users to understand how promise values propagate.

Usually when people write `.catch(err => { throw err; })` they intend to throw the error into the next tick to make sure it's reported, not to add a no-op to the end of their promise chain.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

Promises are widely used, but currently still not well understood.  Including this rule in core will help people learn how to correctly use promises.

**Are you willing to create the rule yourself?**

Yes

**Comment by mysticatea** [comment]
Thank you for this issue.

I have a concern as same as #5760.

**Comment by ilyavolodin** [comment]
I think this should be added to [eslint-plugin-promise](https://github.com/xjamundx/eslint-plugin-promise) instead of the core.

**Comment by kaicataldo** [comment]
@ilyavolodin Agreed, since the plugin exists. Though it might be worth considering including the plugin in core at some point, now that native promises are a thing :)

**Comment by alberto** [comment]
Agreed
